### PR TITLE
Feature/more scantool

### LINKED ISSUE #2479: Very strange behavior of settlers after the transition from version 1.12.2-0.8.8242 to version 1.12.2-0.8.8298
### Minecraft with mods

The settlers stand still. When you click on the "Call employee" button in the building do not respond.
When you click on the "call a resident" button in the town hall calls everyone, but it does not help.
The workers stopped working. It is necessary to re-appoint workers. When appointing workers, they behave strangely. Do not change clothes, for example, when appointing a former shepherd shepherd to the position of a miner, he still stands in farm clothes, does not ask for anything and does nothing. When appointing a former knight again to the position of a knight, he runs around with a bow and not with a sword!

When appointing a courier, a miner (a former shepherd) asks for scissors, an axe and wheat!

### LINKED ISSUE #2480: Player built a hut then their game crashed
pastebin: https://pastebin.com/wTaNTNE8

### Minecolonies version

build 1.12.2-0.8.8242

### Expected behavior

- not crash

### Actual behaviour

- crashed

### Steps to reproduce the problem

1. they said they finished building a hut

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #2480 
Closes #2479

# Changes proposed in this pull request:
- Added a try and catch around that block case this happens in the future again.
- Adds a bounding box to the scan tool.
- Adds a search bar to it as well.

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/commands/generalcommands/RandomTeleportCommand.java [hunk 1]
```diff
@@ -154,7 +154,7 @@ private static void teleportPlayer(final ICommandSender sender, final EntityPlay
 
             final Colony colony = ColonyManager.getClosestColony(sender.getEntityWorld(), tpPos);
             /* Check for a close by colony*/
-            if (BlockPosUtil.getDistance2D(colony.getCenter(), tpPos) < Configurations.gameplay.workingRangeTownHall * 2 + Configurations.gameplay.townHallPadding)
+            if (colony != null && BlockPosUtil.getDistance2D(colony.getCenter(), tpPos) < Configurations.gameplay.workingRangeTownHall * 2 + Configurations.gameplay.townHallPadding)
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/commands/generalcommands/RandomTeleportCommand.java:157]: MAJOR Split this 181 characters long line (which is greater than 180 authorized). [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00103)

File: src/main/java/com/minecolonies/coremod/event/ClientEventHandler.java [hunk 2]
```diff
@@ -73,6 +83,67 @@ public void renderWorldLastEvent(@NotNull final RenderWorldLastEvent event)
...
+            GlStateManager.tryBlendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA,
+                    GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA,
+                    GlStateManager.SourceFactor.ONE,
+                    GlStateManager.DestFactor.ZERO);
+            GlStateManager.glLineWidth(2.0F);
+            GlStateManager.disableTexture2D();
+            GlStateManager.depthMask(false);
+
+            final AxisAlignedBB axisalignedbb = new AxisAlignedBB(x1, y1, z1, x2, y2, z2);
+            RenderGlobal.drawSelectionBoundingBox(axisalignedbb.grow(0.002D).offset(-renderPosX, -renderPosY, -renderPosZ), 1.0F, 1.0F, 1.0F, 1.0F);
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/event/ClientEventHandler.java:140]: MAJOR Assign this magic number 0.002D to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/coremod/items/ItemScanTool.java [hunk 3]
```diff
@@ -7,16 +7,16 @@
 import com.minecolonies.coremod.client.gui.WindowScan;
 import com.minecolonies.coremod.creativetab.ModCreativeTabs;
 import com.minecolonies.coremod.network.messages.SaveScanMessage;
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.entity.Entity;
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/items/ItemScanTool.java]: MINOR Remove this unused import 'net.minecraft.entity.Entity'. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/minecolonies/coremod/items/ItemScanTool.java [hunk 4]
```diff
@@ -7,16 +7,16 @@
...
+import net.minecraft.entity.EntityLivingBase;
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/items/ItemScanTool.java]: MINOR Remove this unused import 'net.minecraft.entity.EntityLivingBase'. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 73 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 62 major
* ![MINOR][MINOR] 6 minor
* ![INFO][INFO] 4 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [WindowScan.java#L323](https://github.com/Minecolonies/minecolonies/blob/d5bc0aca0e15349679934aa284cb9bfd2decac17/src/main/java/com/minecolonies/coremod/client/gui/WindowScan.java#L323): Refactor this code to not nest more than 5 if/for/while/switch/try statements. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS134)
1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/d5bc0aca0e15349679934aa284cb9bfd2decac17/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/d5bc0aca0e15349679934aa284cb9bfd2decac17/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [WindowScan.java#L53](https://github.com/Minecolonies/minecolonies/blob/d5bc0aca0e15349679934aa284cb9bfd2decac17/src/main/java/com/minecolonies/coremod/client/gui/WindowScan.java#L53): Remove this unused "ENTER_KEY" private field. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1068)
1. ![MAJOR][MAJOR] [WindowScan.java#L266](https://github.com/Minecolonies/minecolonies/blob/d5bc0aca0e15349679934aa284cb9bfd2decac17/src/main/java/com/minecolonies/coremod/client/gui/WindowScan.java#L266): The Cyclomatic Complexity of this method "updateResources" is 18 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [WindowScan.java#L266](https://github.com/Minecolonies/minecolonies/blob/d5bc0aca0e15349679934aa284cb9bfd2decac17/src/main/java/com/minecolonies/coremod/client/gui/WindowScan.java#L266): Refactor this method to reduce its Cognitive Complexity from 46 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/d5bc0aca0e15349679934aa284cb9bfd2decac17/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [buildings#](https://github.com/Minecolonies/minecolonies/blob/d5bc0aca0e15349679934aa284cb9bfd2decac17/src/main/java/com/minecolonies/coremod/colony/buildings#): Add a 'package-info.java' file to document the 'buildings' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [permissions#](https://github.com/Minecolonies/minecolonies/blob/d5bc0aca0e15349679934aa284cb9bfd2decac17/src/main/java/com/minecolonies/coremod/colony/permissions#): Add a 'package-info.java' file to document the 'permissions' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [requestable#](https://github.com/Minecolonies/minecolonies/blob/d5bc0aca0e15349679934aa284cb9bfd2decac17/src/main/java/com/minecolonies/coremod/colony/requestable#): Add a 'package-info.java' file to document the 'requestable' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 75 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MAJOR][MAJOR] 63 major
* ![MINOR][MINOR] 5 minor
* ![INFO][INFO] 5 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [WindowScan.java#L323](https://github.com/Minecolonies/minecolonies/blob/57507b6fc80ea5e417e565528369b1b295073704/src/main/java/com/minecolonies/coremod/client/gui/WindowScan.java#L323): Refactor this code to not nest more than 5 if/for/while/switch/try statements. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS134)
1. ![CRITICAL][CRITICAL] [AbstractBuildingWorker.java#L69](https://github.com/Minecolonies/minecolonies/blob/57507b6fc80ea5e417e565528369b1b295073704/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L69): Remove this call from a constructor to the overridable "getBuildingLevel" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1699)
1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/57507b6fc80ea5e417e565528369b1b295073704/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/57507b6fc80ea5e417e565528369b1b295073704/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [WindowScan.java#L53](https://github.com/Minecolonies/minecolonies/blob/57507b6fc80ea5e417e565528369b1b295073704/src/main/java/com/minecolonies/coremod/client/gui/WindowScan.java#L53): Remove this unused "ENTER_KEY" private field. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1068)
1. ![MAJOR][MAJOR] [WindowScan.java#L266](https://github.com/Minecolonies/minecolonies/blob/57507b6fc80ea5e417e565528369b1b295073704/src/main/java/com/minecolonies/coremod/client/gui/WindowScan.java#L266): The Cyclomatic Complexity of this method "updateResources" is 18 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [WindowScan.java#L266](https://github.com/Minecolonies/minecolonies/blob/57507b6fc80ea5e417e565528369b1b295073704/src/main/java/com/minecolonies/coremod/client/gui/WindowScan.java#L266): Refactor this method to reduce its Cognitive Complexity from 46 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/57507b6fc80ea5e417e565528369b1b295073704/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [buildings#](https://github.com/Minecolonies/minecolonies/blob/57507b6fc80ea5e417e565528369b1b295073704/src/main/java/com/minecolonies/coremod/colony/buildings#): Add a 'package-info.java' file to document the 'buildings' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L477](https://github.com/Minecolonies/minecolonies/blob/57507b6fc80ea5e417e565528369b1b295073704/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L477): Reduce this class from 83 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 71 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 62 major
* ![MINOR][MINOR] 4 minor
* ![INFO][INFO] 4 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [WindowScan.java#L323](https://github.com/Minecolonies/minecolonies/blob/1ad3a0fcb02cb86538cac7bb715d91a970607b86/src/main/java/com/minecolonies/coremod/client/gui/WindowScan.java#L323): Refactor this code to not nest more than 5 if/for/while/switch/try statements. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS134)
1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/1ad3a0fcb02cb86538cac7bb715d91a970607b86/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/1ad3a0fcb02cb86538cac7bb715d91a970607b86/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [WindowScan.java#L53](https://github.com/Minecolonies/minecolonies/blob/1ad3a0fcb02cb86538cac7bb715d91a970607b86/src/main/java/com/minecolonies/coremod/client/gui/WindowScan.java#L53): Remove this unused "ENTER_KEY" private field. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1068)
1. ![MAJOR][MAJOR] [WindowScan.java#L266](https://github.com/Minecolonies/minecolonies/blob/1ad3a0fcb02cb86538cac7bb715d91a970607b86/src/main/java/com/minecolonies/coremod/client/gui/WindowScan.java#L266): The Cyclomatic Complexity of this method "updateResources" is 18 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [WindowScan.java#L266](https://github.com/Minecolonies/minecolonies/blob/1ad3a0fcb02cb86538cac7bb715d91a970607b86/src/main/java/com/minecolonies/coremod/client/gui/WindowScan.java#L266): Refactor this method to reduce its Cognitive Complexity from 46 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/1ad3a0fcb02cb86538cac7bb715d91a970607b86/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [buildings#](https://github.com/Minecolonies/minecolonies/blob/1ad3a0fcb02cb86538cac7bb715d91a970607b86/src/main/java/com/minecolonies/coremod/colony/buildings#): Add a 'package-info.java' file to document the 'buildings' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [permissions#](https://github.com/Minecolonies/minecolonies/blob/1ad3a0fcb02cb86538cac7bb715d91a970607b86/src/main/java/com/minecolonies/coremod/colony/permissions#): Add a 'package-info.java' file to document the 'permissions' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [requestable#](https://github.com/Minecolonies/minecolonies/blob/1ad3a0fcb02cb86538cac7bb715d91a970607b86/src/main/java/com/minecolonies/coremod/colony/requestable#): Add a 'package-info.java' file to document the 'requestable' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Review by marchermans — APPROVED** [review]

File: src/main/java/com/minecolonies/coremod/commands/generalcommands/RandomTeleportCommand.java (hunk 1 above)
Comment (kreezxil) [reply src/main/java/com/minecolonies/coremod/commands/generalcommands/RandomTeleportCommand.java:157]: I still don't understand why you would square the players/admin Config setting for workingRangeTownHall. I understand squaring coordinates and subtracting the squares to then get the squareroot to get a distance between the two. However, the Config value is already a distance, so why are you squaring it?

**Pull request merged**